### PR TITLE
[Fix] Avoid infinite GPU waiting in dist training

### PR DESCRIPTION
**Pull request by fingertap**
See #6495 for details. Here I just add an assertion before reducing the log_vars to ensure every GPU has the same length of log_vars to prevent infinite waiting.

### CODE REVIEW

File: mmdet/models/detectors/base.py [hunk 1]
```diff
@@ -198,6 +198,13 @@ def _parse_losses(self, losses):
         loss = sum(_value for _key, _value in log_vars.items()
                    if 'loss' in _key)
 
+        # If the loss_vars has different length, GPUs will wait infinitely
+        if dist.is_available() and dist.is_initialized():
```
Comment (hhaAndroid) [inline mmdet/models/detectors/base.py:202]: You can directly use `reduce_mean` in mmdet/core/utils/dist_utils.py

File: mmdet/models/detectors/base.py [hunk 2]
```diff
@@ -198,6 +198,13 @@ def _parse_losses(self, losses):
...
+            log_var_length = torch.tensor(len(log_vars), device=loss.device)
+            dist.all_reduce(log_var_length)
+            assert log_var_length == len(log_vars) * dist.get_world_size(), \
```
Comment (hhaAndroid) [inline mmdet/models/detectors/base.py:205]: Once an error occurs, print out all the keys in each process to facilitate troubleshooting.

Comment (fingertap) [reply mmdet/models/detectors/base.py:205]: does mmdet provide any thread-safe print utilities?

File: mmdet/models/detectors/base.py (hunk 1 above)
Comment (fingertap) [reply mmdet/models/detectors/base.py:202]: Both ways are good

File: mmdet/models/detectors/base.py (hunk 2 above)
Comment (hhaAndroid) [reply mmdet/models/detectors/base.py:205]: No need to think about it, only need to be printed after asset, and keys and rank need to be printed at the same time.

File: mmdet/models/detectors/base.py (hunk 1 above)
Comment (hhaAndroid) [reply mmdet/models/detectors/base.py:202]: It's better to be able to write uniformly.

Comment (fingertap) [reply mmdet/models/detectors/base.py:202]: The communication in the [following lines](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/detectors/base.py#L206) are also through `torch.dist`. So I just use a similar way to do this. If you want to use the utilities from mmdet, I suggest a more complete style check.

File: mmdet/models/detectors/base.py (hunk 2 above)
Comment (fingertap) [reply mmdet/models/detectors/base.py:205]: Also, how can other GPUs know that this GPU raises error?

File: mmdet/models/detectors/base.py (hunk 1 above)
Comment (hhaAndroid) [reply mmdet/models/detectors/base.py:202]: Okay

File: mmdet/models/detectors/base.py (hunk 2 above)
Comment (fingertap) [reply mmdet/models/detectors/base.py:205]: It seems that if we want to inform all GPUs about the assertion error, we have to do one more communication among GPUs. If this overhead is Okay, I will add it to this PR.

Comment (hhaAndroid) [reply mmdet/models/detectors/base.py:205]: There is no need to distinguish which GPU is wrong, just print all the keys, the user can check the output to determine which key is wrong

`assert Flase, f'{...get_rank(), log_val.keys(), len(log_val.keys())...}'`

**Comment by fingertap** [comment]
Before I could merge, the upstream master need to fix the BC bug as in mmseg.

File: mmdet/models/detectors/base.py (hunk 2 above)
Comment (fingertap) [reply mmdet/models/detectors/base.py:205]: Yes, you are right. At least two GPUs will raise exception, so users can compare the error messages.

Comment (hhaAndroid) [reply mmdet/models/detectors/base.py:205]: Yes

**Comment by RangiLyu** [comment]
Hi, we have fixed the compatibility with MMCV in v2.18.1. Now the CI can work normally. Please pull the latest master and re-run the unit tests.

**Comment by fingertap** [comment]
> Hi, we have fixed the compatibility with MMCV in v2.18.1. Now the CI can work normally. Please pull the latest master and re-run the unit tests.

Merged.

**Comment by RangiLyu** [comment]
Lint failed. Please pip install pre-commit and then run  `pre-commit run --all-files` to fix.

**Review by RangiLyu — APPROVED** [review]

File: mmdet/models/detectors/base.py [hunk 3]
```diff
@@ -198,6 +198,16 @@ def _parse_losses(self, losses):
         loss = sum(_value for _key, _value in log_vars.items()
                    if 'loss' in _key)
 
+        # If the loss_vars has different length, GPUs will wait infinitely
```
Comment (ZwwWayne) [inline mmdet/models/detectors/base.py:201]: suggest to use rank, word_size = get_dist_info(), then use word_size to decide whether to synchronize

File: mmdet/models/detectors/base.py [hunk 4]
```diff
@@ -198,6 +198,16 @@ def _parse_losses(self, losses):
...
+        if dist.is_available() and dist.is_initialized():
+            log_var_length = torch.tensor(len(log_vars), device=loss.device)
+            dist.all_reduce(log_var_length)
```
Comment (ZwwWayne) [inline mmdet/models/detectors/base.py:204]: You may consider reduce_mean in mmdet, and use assert log_var_length == len(log_vars)

File: mmdet/models/detectors/base.py (hunk 3 above)
Comment (fingertap) [reply mmdet/models/detectors/base.py:201]: No need to broadcast the error. At least two GPUs will have their len(log_vars) different from the mean. So users can compare the error logs to determine the missing loss terms.

**Review by hhaAndroid — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Comment by OpenMMLab-Coodinator** [comment]
Hi Han！First of all, we want to express our gratitude for your significant PR in the MMDet project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you are Chinese or have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤

**Comment by fingertap** [comment]
Thanks for your invitation.
